Add improvements to modules, services, and contributions pages (#889)
gruntwork-io/docs: _docs-sources/library/overview/modules.md

# What is a module?

> **Modules are reusable "infrastructure building blocks" that describe how to deploy and manage a specific piece of infrastructure, such as a VPC, ECS cluster, or Auto Scaling Group.**

Reviewer: ```suggestion
Modules are reusable "infrastructure building blocks" that describe how to deploy and manage a specific piece of infrastructure, such as a VPC, ECS cluster, or Auto Scaling Group.
```

Most modules are written in Terraform and define several AWS resources.

## Example

Let’s look at an example module. The [rds module](/reference/modules/terraform-aws-data-storage/rds/) is a Terraform module that creates an RDS database, the IAM roles needed to operate that database, optional read replicas, database subnet groups, and the relevant security groups.

> The module deploys a key element of an overall RDS deployment, but it's not a _complete_ RDS deployment. That's because the `rds` module does not include backup policies using AWS Backup (for disaster recovery), or RDS Proxy (to pool database connections), or CloudWatch alarms (to alert you when something goes wrong). These missing pieces are best thought of as building block modules themselves! Indeed, `backup-plan`, `backup-vault`, and `rds-proxy` can all be used in combination with the `rds` module.

Reviewer: ```suggestion
The module deploys a key element of an overall RDS deployment, but it's not a _complete_ RDS deployment. That's because the `rds` module does not include backup policies using AWS Backup (for disaster recovery), or RDS Proxy (to pool database connections), or CloudWatch alarms (to alert you when something goes wrong). These missing pieces are best thought of as building block modules themselves. Gruntwork has modules for `backup-plan`, `backup-vault`, and `rds-proxy` can all be used in combination with the `rds` module.
```

To see how Gruntwork gives you an off-the-shelf overall deployment with all the elements included, see [What is a service module?](./services.md).

## Modules are optimized for control

> A "building block" module is designed to be small, narrow in scope, and highly reusable. In other words, building block modules give you _control_, but they may not give you _convenience_. That is, you can use the building block modules for all kinds of use cases (high control), but if you want to deploy a complete piece of infrastructure, you still have to do the work of assembling the right building blocks (low convenience).

Reviewer: ```suggestion
A module is designed to be small, narrow in scope, and highly reusable, like a building block. Modules give you _control_, but they may not give you _convenience_. You can use the building block modules for all kinds of use cases (high control), but if you want to deploy a complete piece of infrastructure, you still have to do the work of assembling the right modules (low convenience).
```

"building block" doesn't resonate with me. I see what you're aiming for but I find it a bit confusing to use it to prefix module, rather than just as a device to build a mental model.

Author: Fair point; I've seen that terminology resonate on customer calls, but I probably overdid it here. Feel free to open a PR to further prune the usage of that term if after this PR is merged you think we're still overdoing it.

To learn how you can optimize for convenience, see [What is a service module?](./services.md).

To learn more about the overall thought process behind building block modules versus service modules, see [Introducing: The Gruntwork Module, Service, and Architecture Catalogs](https://blog.gruntwork.io/introducing-the-gruntwork-module-service-and-architecture-catalogs-eb3a21b99f70).

## When to use a building block module

> Building block modules are fairly generic by design, so you don’t typically deploy a single building block module directly. Instead, you write code that combines the building block modules you need for a specific use case.

Reviewer: ```suggestion
Building block modules are fairly generic by design, so you won't typically deploy a single building block module directly. Instead, you write code that combines the building block modules you need for a specific use case.
```

For example, one building block module might deploy the control plane for Kubernetes and a separate building block module could deploy worker nodes; you may need to combine both modules together to deploy a Kubernetes cluster.

We recommend our [Service Catalog](/library/overview/services/) for common use cases, but our full module catalog is available if you have a more complex use case.

## Where to find the building block modules

The module catalog features over 250 "building block" modules spanning three major use cases:

1. AWS foundations

> 2. Run my app

Reviewer: ```suggestion
2. Running applications
```

3. Store my data

> Each of these use cases covers one or more "SME topics" such as AWS account management, VPC/Networking, EKS, ECS, and RDS. SME topics are a first-class concept within Gruntwork, but do not yet have much visibility in the product itself.

Reviewer: ```suggestion
Each of these use cases covers one or more Subject Matter Expert (SME) topics such as AWS account management, VPC/Networking, EKS, ECS, and RDS. SME topics are a first-class concept within Gruntwork, but do not have much visibility in the product itself at this time.
```

Always spell out an acronym the first time you use it on a page.

1. `test`: Automated tests for the code in modules and examples.

We follow Hashicorp's [Standard Model Structure](https://developer.hashicorp.com/terraform/language/modules/develop/structure) for our files (`main.tf`, `variables.tf`, `outputs.tf`). In the `variables.tf` file we always put the required variables at the top of the file, followed by the optional variables. Although there are often a lot of ways to configure our modules, we set reasonable defaults and try to minimize the effort required to configure the modules to the most common use cases.

## How modules are updated

> Gruntwork brings together AWS and Terraform experts around the world who track updates from AWS, Terraform, and the DevOps community at large, along with requests from the Gruntwork customer community. We translate the most important of these updates into new features, new optimizations, and ultimately new releases.

Reviewer: This describes an abstract concept of how we update. We should link to the page that describes it in detail. `/library/overview/modules#how-modules-are-updated`

Author: Fair point, though I don't see the page you're referring to? The path you listed refers to this section, and I couldn't find any other pages so will leave for now.

> Check out [Gruntwork releases](/guides/stay-up-to-date/#gruntwork-releases) for a comprehensive listing of all the updates.

Reviewer: ```suggestion
Refer to [Gruntwork releases](/guides/stay-up-to-date/#gruntwork-releases) for a comprehensive listing of all the updates.
```

Author: Fixed in 9fcccdcc5de1280991cd3510e6ff81731a20c6dd!
